File: Spark operations (e.g., transformations and actions) and MapReduce/Scala/task3.scala
import java.io._
import org.apache.spark.Partitioner
import org.apache.spark.sql.SparkSession
import org.json4s._
import org.json4s.jackson.JsonMethods._


  object task3 {
    def main(args: Array[String]): Unit = {

      if (args.length != 5) {
        println("Please enter all the required agruments")
        sys.exit(1)
      }

      val ss = SparkSession
        .builder()
        .appName(name = "task3")
        .config("spark.master", "local[*]")
        .getOrCreate()

      val sc = ss.sparkContext
      sc.setLogLevel("ERROR")

      val inputFile = args(0)
      val rdd = sc.textFile(inputFile)
      val dataRdd = rdd.map(string => parse(string).values.asInstanceOf[Map[String, String]])

      if(args(2) == "default"){
        val n_partitions = dataRdd.map(x=>(x("business_id"),1)).groupByKey().map(x=>(x._1,x._2.size)).filter(x=> x._2>args(4).toInt).getNumPartitions
        val n_items = dataRdd.map(x=>(x("business_id"),1)).groupByKey().map(x=>(x._1,x._2.size)).filter(x=> x._2>args(4).toInt).glom().map(_.length).collect()
        val result = dataRdd.map(x=>(x("business_id"),1)).groupByKey().map(x=>(x._1,x._2.size)).filter(x=> x._2>args(4).toInt).collect()

        val output_file = new File(args(1))
        val output_writer = new BufferedWriter(new FileWriter(output_file))
        output_writer.write("{"+ "\"n_partitions\":"+n_partitions.toInt+","+"\"n_items\":"+n_items.toList.toString().replaceAll("List","").replaceAll("\\(", "[").replaceAll("\\)", "]")+","+"\"result\":"+result.toList.toString().replaceAll("List","").replaceAll("\\(", "[").replaceAll("\\)", "]")+"}")
        output_writer.close()
      }


      if (args(2) == "customized")
      {
        class customizedPartition(totalPartition: Int) extends Partitioner{
          override def numPartitions: Int = totalPartition
          override def getPartition(key: Any): Int = key match {
            case null => 0
            case s:String =>
              val rawMod = s.takeRight(1).hashCode % numPartitions
              rawMod + (if (rawMod < 0) s.takeRight(1).hashCode else 0)
          }
        }

        val patitioned = dataRdd.map(x=>(x("business_id"),1)).partitionBy(new customizedPartition(args(3).toInt)).persist()
        val reduced_data = patitioned.reduceByKey((x,y)=>x+y).filter(x=> x._2 > args(4).toInt)

        val n_items = reduced_data.glom().map(_.length).collect()
        val result = reduced_data.collect()

        val output_file = new File(args(1))
        val output_writer = new BufferedWriter(new FileWriter(output_file))
        output_writer.write("{"+ "\"n_partitions\":"+args(3).toInt+","+"\"n_items\":"+n_items.toList.toString().replaceAll("List","").replaceAll("\\(", "[").replaceAll("\\)", "]")+","+"\"result\":"+result.toList.toString().replaceAll("List","").replaceAll("\\(", "[").replaceAll("\\)", "]")+"}")
        output_writer.close()
      }
      sc.stop()
    }
  }

File: Spark operations (e.g., transformations and actions) and MapReduce/Python/task3.py
from pyspark import SparkContext
import json
import sys


def sub_tasks(review_rdd, output_file, partition_type, n_part, n):
    if str(partition_type) == "default":
        n_partitions = review_rdd.map(lambda x: (x['business_id'], 1)).groupByKey().map(lambda x: (x[0], len(x[1]))) \
            .filter(lambda x: x[1] > int(n)).getNumPartitions()

        n_items = review_rdd.map(lambda x: (x['business_id'], 1)).groupByKey().map(lambda x: (x[0], len(x[1]))) \
            .filter(lambda x: x[1] > int(n)).glom().map(len).collect()

        result = review_rdd.map(lambda x: (x['business_id'], 1)).groupByKey().map(lambda x: (x[0], len(x[1]))) \
            .filter(lambda x: x[1] > int(n)).collect()

        output = {
            "n_partitions": n_partitions,
            "n_items": n_items,
            "result": result,
        }

        with open(str(output_file), "w+") as fileout:
            json.dump(output, fileout, indent=4)

    elif str(partition_type) == "customized":
        r = review_rdd.map(lambda x: (x['business_id'], 1)).partitionBy(int(n_part),
                                                                        lambda x: ord(x[-1])).reduceByKey(lambda x, y: x + y).filter(lambda x: x[1] > int(n)).persist()
        n_items = r.glom().map(len).collect()

        result = r.collect()

        output = {
            "n_partitions": int(n_part),
            "n_items": n_items,
            "result": result,
        }

        with open(str(output_file), "w+") as fileout:
            json.dump(output, fileout, indent=4)


def main():
    if len(sys.argv) != 6:
        print("Please enter all the required agruments")
        exit(-1)

    sc = SparkContext('local[*]', 'task3')
    review_rdd = sc.textFile(str(sys.argv[1])).map(lambda x: json.loads(x))

    sub_tasks(review_rdd, sys.argv[2], sys.argv[3], sys.argv[4], sys.argv[5])

    sc.stop()


if __name__ == "__main__": main()
